### ISSUE TITLE
Error when no strict refs in effect

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6077]
Migrated from [rt.perl.org#18479](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=18479) (status was 'resolved')

Searchable as RT18479$

**Comment by p5pRT** [comment]
### From rov_f@web.de

There is an error in Perl 5\.6\.1 when no strict refs is used within an eval\.\.\.\. 
See the file report\.txt in the enclosed tar file for an exact description\.

Ronald
\-\- 
Ronald Otto Valentin Fischer \<rov\_f@&#8203;web\.de>
http&#8203;://profiles\.yahoo\.com/ronny\_fischer/
\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_
Viren? Wir wissen nicht was Ihr Arzt empfiehlt\. Wir empfehlen den
Virencheck für Dateianhänge\! http&#8203;://freemail\.web\.de/features/?mc=021159

**Comment by p5pRT** [comment]
### From rov_f@web.de

[bugreport.tar](https://rt-archive.perl.org/perl5/Ticket/Attachment/41865/33652/bugreport.tar)

**Comment by p5pRT** [comment]
### From @mjdominus


The problem is not exatly as Mr\. Fischer describes\.  Here is a simpler
demonstration&#8203;:

&nbsp;        \#\!/usr/bin/perl
&nbsp;        use strict;
&nbsp;        my $prog1= qq\{sub Du \{                \} print STDERR "hello\\n"; \};
&nbsp;        my $prog2= qq\{sub Du \{ no strict refs \} print STDERR "hello\\n"; \};


&nbsp;        print "prog1&#8203;:\\n";
&nbsp;        eval $prog1;
&nbsp;        print "aborted&#8203;: $@&#8203;\\n" if $@&#8203; ne '';
&nbsp;        print "terminated\\n";
&nbsp;        print "\\n";

&nbsp;        print "prog2&#8203;:\\n";
&nbsp;        eval $prog2;
&nbsp;        print "aborted&#8203;: $@&#8203;\\n" if $@&#8203; ne '';
&nbsp;        print "terminated\\n";


The output is&#8203;:

&nbsp;        prog1&#8203;:
&nbsp;        hello
&nbsp;        terminated

&nbsp;        prog2&#8203;:
&nbsp;        terminated

Note that 'hello' is missing from prog2\.

The problem here is that the 'use strict' in the main program enable
'strict subs'\, which is inherited by the 'evals'\.  $prog2 fails to
compile because it contains the bareword 'refs'\.  When Perl tries to
compile $prog2\, there is a strict\-subs failure\.

The real problem is that afer the eval $prog2 fails\, $@&#8203; is not set\.
It should contain something like 

&nbsp;        Bareword "refs" not allowed while "strict subs" in use

but instead\, it's empty\.

**Comment by p5pRT** [comment]
### From @rgs

Mark Jason Dominus wrote&#8203;:
> 
> The real problem is that afer the eval $prog2 fails\, $@&#8203; is not set\.

Cut\-down test case :

\#\!/usr/bin/perl
use strict 'subs';
eval 'BEGIN\{x\}';
print "\# $@&#8203;\\n";
print $@&#8203; ? "ok\\n" : "not ok\\n";
\_\_END\_\_

**Comment by p5pRT** [comment]
### From @andk

>>>>> On Mon\, 18 Nov 2002 18&#8203;:08&#8203;:07 \+0100\, Rafael Garcia\-Suarez \<rgarciasuarez@&#8203;free\.fr> said&#8203;:

 rg> Mark Jason Dominus wrote&#8203;:
 >> 
 >> The real problem is that afer the eval $prog2 fails\, $@&#8203; is not set\.

 rg> Cut\-down test case :


\-\-\-\-Program\-\-\-\-
\#\!/usr/bin/perl
use strict 'subs';
eval 'BEGIN\{x\}';
print "\# $@&#8203;\\n";
print $@&#8203; ? "ok\\n" : "not ok\\n";

\-\-\-\-Output of /usr/local/perl\-5\.005\_56@&#8203;3446/bin/perl\-\-\-\-
Bareword "x" not allowed while "strict subs" in use at \(eval 1\) line 1\.
\# BEGIN not safe after errors\-\-compilation aborted at \(eval 1\) line 1\.

ok

\-\-\-\-EOF \($?='0'\)\-\-\-\-
\-\-\-\-Output of /usr/local/perl\-5\.005\_56@&#8203;3447/bin/perl\-\-\-\-
Bareword "x" not allowed while "strict subs" in use at \(eval 1\) line 1\.
\# 
not ok

\-\-\-\-EOF \($?='0'\)\-\-\-\-

\-\- 
andreas

**Comment by p5pRT** [comment]
### From @eserte

Rafael Garcia\-Suarez \<rgarciasuarez@&#8203;free\.fr> writes&#8203;:

> Mark Jason Dominus wrote&#8203;:
> > 
> > The real problem is that afer the eval $prog2 fails\, $@&#8203; is not set\.
> 
> Cut\-down test case :
> 
> \#\!/usr/bin/perl
> use strict 'subs';
> eval 'BEGIN\{x\}';
> print "\# $@&#8203;\\n";
> print $@&#8203; ? "ok\\n" : "not ok\\n";
> \_\_END\_\_
> 

Here's a crude fix\. Or better&#8203;: just the point in code where the fix
should probably go\.

Regards\,
&nbsp;    Slaven

<details><summary>Inline Patch</summary>

```diff
--- bleedperl/pp_ctl.c	Thu Oct 31 00:42:48 2002
+++ bleedperl2/pp_ctl.c	Mon Nov 18 22:41:25 2002
@@ -2732,6 +2732,12 @@ S_doeval(pTHX_ int gimme, OP** startop)
 	    Perl_croak(aTHX_ "%sCompilation failed in regexp",
 		       (*msg ? msg : "Unknown error\n"));
 	}
+	else {
+	    char* msg = SvPVx(ERRSV, n_a);
+	    if (!*msg) {
+	        sv_setpv(ERRSV, "Compilation error");
+	    }
+	}
 	RETPUSHUNDEF;
     }
     CopLINE_set(&PL_compiling, 0);

-- 
```
</details>

Slaven Rezic \- slaven\.rezic@&#8203;berlin\.de

&nbsp;    tktimex \- project time manager
&nbsp;    http&#8203;://sourceforge\.net/projects/ptktools/

**Comment by p5pRT** [comment]
### From @hvds

Slaven Rezic \<slaven\.rezic@&#8203;berlin\.de> wrote&#8203;:
:Rafael Garcia\-Suarez \<rgarciasuarez@&#8203;free\.fr> writes&#8203;:
:> Cut\-down test case :
:> 
:> \#\!/usr/bin/perl
:> use strict 'subs';
:> eval 'BEGIN\{x\}';
:> print "\# $@&#8203;\\n";
:> print $@&#8203; ? "ok\\n" : "not ok\\n";
:> \_\_END\_\_
:> 
:
:Here's a crude fix\. Or better&#8203;: just the point in code where the fix
:should probably go\.

Thanks\, crude fix applied as \#18176\.

So where has the \_real\_ $@&#8203; gone \- did we just restore over it?
Or did the message never get to $@&#8203; in the first place?

Hugo

**Comment by p5pRT** [comment]
@cwest - Status changed from 'new' to 'resolved'
